### PR TITLE
Shelley: only keep a compact genesis config in memory

### PR DESCRIPTION
**Pull request by mrBliss**
Some fields in the Shelley genesis config are only needed at start-up. There's
no point in keeping them in memory during the rest of the execution. In some
testing setups, e.g., for transaction benchmarking, the genesis config can be
quite large.

This also means that the `GetGenesisConfig` query (#2694) would only return the
compact genesis config, not the original full one.

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Config.hs [hunk 1]
```diff
@@ -90,3 +93,26 @@ data instance StorageConfig (ShelleyBlock era) = ShelleyStorageConfig {
...
   deriving (Generic, NoThunks)
+
+{-------------------------------------------------------------------------------
+  Compact genesis
+-------------------------------------------------------------------------------}
+
+-- | Compact variant of 'SL.ShelleyGenesis' with some fields erased that are
+-- only used on start-up and that should not be kept in memory forever.
+--
+-- For example, the 'sgInitialFunds' field.
```
Comment (mrBliss) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Config.hs:104]: State which two fields are cleared.

**Review by edsko — APPROVED** [review]
Not terribly keen on having this "record with some fields erased" approach. Surely it would be cleaner to have a different type that makes it clear that we don't have values for these fields?

**Comment by mrBliss** [comment]
> Not terribly keen on having this "record with some fields erased" approach. Surely it would be cleaner to have a different type that makes it clear that we don't have values for these fields?

That would indeed be cleaner. I guess `ShelleyGenesis` shouldn't have contained these fields in the first place. But it does, and they're part of the hash of the file, and that's part of the chain, ...

Also, we don't want consensus to be in the business of defining genesis configs for a ledger.

**Comment by mrBliss** [comment]
bors merge

**Pull request merged**